### PR TITLE
[Backport] [2.x] [Concurrent Segment Search]: Add new query profile collector fields with concurrent search execution (#7898)

### LINKED ISSUE #7355: [Concurrent Segment Search] Query profile stats with concurrent execution of query phase
With current model of concurrent search, seems like we add time take by different slices to reflect the time taken by a collector. This may not be the true reflection of time taken by that collector. We should explore and see if instead it would make sense to expose max/min/average latency across different slices execution of a collector. For more context see https://github.com/opensearch-project/OpenSearch/pull/1500

### PR DESCRIPTION
**Pull request by ticheng-aws**
(cherry picked from commit c58dc1ade9096ab70e22024e91eecd79fe2416e8)

### Description
Backport of https://github.com/opensearch-project/OpenSearch/pull/7898 to `2.x` branch.

To show high-level concurrent execution details, we bring `min_slice_time_in_nanos`, `max_slice_time_in_nanos`, and `slice_count` etc. fields into the query response. With these new fields, users can quickly determine query performance and skew issues.

#### Definition of existing collector fields
`name`: A more friendly representation of the collector's class name.
`reason`: A "hint" to help provide some context about this collector.
`time_in_nanos`: The total elapsed time for this collector. In the concurrent search case, this is the total time across all slices (i.e. `max(slice_end_time) - min(slice_start_time)`)
`children`: A list of children collectors "embedded" inside this collector.

#### Definition of new collector fields with concurrent execution
`reduce_time_in_nanos`: The elapsed time in the reduce phase.
`min_slice_time_in_nanos`: The minimum slice time for this collector manager.
`max_slice_time_in_nanos`: The maximum slice time for this collector manager.
`avg_slice_time_in_nanos`: The average slice time for this collector manager.
`slice_count`: The segment slice count for this collector manager.

### Related Issues
Resolves #7355
Doc update issue: https://github.com/opensearch-project/documentation-website/issues/4281

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java [hunk 1]
```diff
@@ -113,6 +164,19 @@ public CollectorResult(StreamInput in) throws IOException {
             CollectorResult child = new CollectorResult(in);
             this.children.add(child);
         }
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java:167]:
```suggestion
        if (in.getVersion().onOrAfter(Version.V_2_9_0)) {
```

File: server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java [hunk 2]
```diff
@@ -124,31 +188,73 @@ public void writeTo(StreamOutput out) throws IOException {
         for (CollectorResult child : children) {
             child.writeTo(out);
         }
+        if (out.getVersion().onOrAfter(Version.V_3_0_0)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java:191]:
```suggestion
        if (out.getVersion().onOrAfter(Version.V_2_9_0)) {
```

**Review by reta — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java (hunk 1 above)
Comment (ticheng-aws) [reply server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java:167]: Thanks @reta for the change request. Our target for the general availability of concurrent search is version `2.10`. We may need to wait until the `2.10` branch and version name are created.

Comment (ticheng-aws) [reply server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java]: Thanks for the clarification, @reta.

**Review by reta — APPROVED** [review]

File: server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java [hunk 3]
```diff
@@ -113,6 +164,19 @@ public CollectorResult(StreamInput in) throws IOException {
             CollectorResult child = new CollectorResult(in);
             this.children.add(child);
         }
+        if (in.getVersion().onOrAfter(Version.V_2_9_0)) {
```
[the remark this answers is not in the document]
Comment (ticheng-aws) [reply server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java:167]: Sure, let me work on this.

[the remark this answers is not in the document]
Comment (ticheng-aws) [reply server/src/main/java/org/opensearch/search/profile/query/CollectorResult.java:167]: Hi @reta, please find the PR here https://github.com/opensearch-project/OpenSearch/pull/7979, and feel free to let me know if you have any questions.

**Pull request merged**